class-two-factor-core.php

```diff
@@ -797,14 +812,24 @@ public static function verify_login_nonce( $user_id, $nonce ) {
 			return false;
 		}
 
-		if ( hash_equals( $login_nonce['key'], self::hash_login_nonce( $nonce ) ) && time() < $login_nonce['expiration'] ) {
-			return true;
+		$unverified_nonce = array(
+			'user_id'    => $user_id,
+			'expiration' => $login_nonce['expiration'],
+			'key'        => $nonce,
+		);
+
+		$unverified_hash = self::hash_login_nonce( $unverified_nonce );
+		$hashes_match    = $unverified_hash && hash_equals( $login_nonce['key'], $unverified_hash );
+
+		$verified = false;
+		if ( $hashes_match && time() < $login_nonce['expiration'] ) {
+			$verified = true;
 		}
 
-		// Require a fresh nonce if verification fails.
+		// Require a fresh nonce if verification fails, and don't allow valid nonces to be reused.
```

Comment:
Valid nonces are already being purged here if I'm not mistaken:

https://github.com/WordPress/two-factor/blob/79c364169e4e7fc430c4ed05c9cc81402955785c/class-two-factor-core.php#L869

So this was specifically dealing with the invalid nonces as it wouldn't reach that point during the invalid logic flow.

Reply:
Ah, you're right, I was looking for that in the wrong place. I'll update it. Thanks!

Reply:
I removed that in 5f37351, but kept the part where the test is renamed for accuracy.

It doesn't look like `login_form_validate_2fa()` is testable, because it's not possible to mock `INPUT_POST`, but that might be something we could test in the future w/ e2e tests. #427 might bring in some from `sjinks/wp-two-factor-provider-webauthn`